File: datacollector/src/main/java/com/example/leo/datacollector/datacollection/DataCollectorService.java
package com.example.leo.datacollector.datacollection;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.hardware.SensorManager;
import android.hardware.display.DisplayManager;
import android.location.Location;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.PowerManager;
import android.preference.PreferenceManager;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;
import android.view.Display;

import com.example.leo.datacollector.R;
import com.example.leo.datacollector.application.DataCollectorApplication;
import com.example.leo.datacollector.database.SqliteDatabase;
import com.example.leo.datacollector.models.SensorDataSet;
import com.example.leo.datacollector.datacollection.sensors.AmbientSound;
import com.example.leo.datacollector.datacollection.sensors.AmbientSoundListener;
import com.example.leo.datacollector.datacollection.sensors.FourSquareListener;
import com.example.leo.datacollector.datacollection.sensors.FoursquareCaller;
import com.example.leo.datacollector.datacollection.sensors.GoogleFitness;
import com.example.leo.datacollector.datacollection.sensors.GoogleFitnessListener;
import com.example.leo.datacollector.datacollection.sensors.GooglePlacesCaller;
import com.example.leo.datacollector.datacollection.sensors.GooglePlacesListener;
import com.example.leo.datacollector.datacollection.sensors.MyActivity;
import com.example.leo.datacollector.datacollection.sensors.MyActivityListener;
import com.example.leo.datacollector.datacollection.sensors.MyEnvironmentSensor;
import com.example.leo.datacollector.datacollection.sensors.MyEnvironmentSensorListener;
import com.example.leo.datacollector.datacollection.sensors.MyLocation;
import com.example.leo.datacollector.datacollection.sensors.MyLocationListener;
import com.example.leo.datacollector.datacollection.sensors.MyMotion;
import com.example.leo.datacollector.datacollection.sensors.MyMotionListener;
import com.example.leo.datacollector.datacollection.sensors.WeatherCaller;
import com.example.leo.datacollector.datacollection.sensors.WeatherCallerListener;
import com.example.leo.datacollector.services.ActivitiesIntentService;
import com.google.android.gms.location.DetectedActivity;

import org.threeten.bp.LocalDateTime;

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import static com.example.leo.datacollector.activityRecording.RecordingActivityKt.RECORDING_ID;
import static com.example.leo.datacollector.utils.ConstantsKt.IS_RECORDING;
import static com.example.leo.datacollector.utils.ConstantsKt.START_RECORDING;
import static com.example.leo.datacollector.utils.ConstantsKt.STOP_RECORDING;
import static com.example.leo.datacollector.utils.UtilsKt.averageDequeue;

public class DataCollectorService extends Service implements MyLocationListener,
        MyMotionListener, FourSquareListener, MyActivityListener,
        MyEnvironmentSensorListener, GooglePlacesListener, AmbientSoundListener,
        GoogleFitnessListener, WeatherCallerListener {

    public static final String TAG = "DataCollectorService";
    public static final int notificationID = 1001;

    private static final int secondsUntilNextUpdate = 5;

    private final int numberOfSamples = 60 / secondsUntilNextUpdate * 5;
    private int sample = 0;
    //rolling buffer of sensorData
    private SensorDataSet[] sensorDataBuffer = new SensorDataSet[numberOfSamples];

    private static final boolean useGooglePlaces = false;
    public static final String ACTIVITY = "activity";
    public static final String TYPE = "type";
    public static final String MINUTES = "minutes";
    public static final String SNACK = "snack";
    public static final String DELETED = "deleted";
    private int weatherUpdateCnt = 0;

    MyMotion myMotion;
    MyLocation myLocation;
    MyActivity myActivity;
    FoursquareCaller foursquareCaller;
    GooglePlacesCaller googlePlacesCaller;
    Location currentLocation;
    MyEnvironmentSensor myEnvironmentSensor;
    AmbientSound ambientSound;
    GoogleFitness googleFitness;
    WeatherCaller weatherCaller;

    double currentLatitude, currentLongitude, currentAccurate, currentAmbientSound;
    boolean ifLocationChanged, isRunning, isPreScreenOn, isCurrentScreenOn;
    String preWifiName, currentWifiName, prePlaceName, currentPlaceName, currentLabel,
            currentWeatherCondition;
    private DetectedActivity preActivity = new DetectedActivity(DetectedActivity.UNKNOWN, 0);
    private DetectedActivity currentActivity = new DetectedActivity(DetectedActivity.UNKNOWN, 0);
    float preAmbientLight, currentAmbientLight, currentTemperature;
    private long currentWeatherId = -1;
    long preSteps, currentSteps;
    private String userName;
    private SqliteDatabase db;
    private int recordingId = -1;

    public DataCollectorService() {

    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        db = SqliteDatabase.Companion.getInstance(getApplicationContext());
        currentWeatherId = db.getLatestWeather().id;
        Log.d(TAG, "service started");
        if (!this.isRunning) {
            startDataCollection();
        }
        if (intent != null && intent.getAction() != null)
            switch (intent.getAction()) {
                case (START_RECORDING): {
                    recordingId = intent.getIntExtra(RECORDING_ID, -1);
                    Intent answer = new Intent(START_RECORDING);
                    answer.putExtra(RECORDING_ID, recordingId);
                    LocalBroadcastManager.getInstance(this).sendBroadcast(answer);
                    break;
                }
                case (STOP_RECORDING): {
                    recordingId = -1;
                    Intent answer = new Intent(STOP_RECORDING);
                    LocalBroadcastManager.getInstance(this).sendBroadcast(answer);
                    break;
                }
                case (IS_RECORDING): {
                    Intent answer = new Intent(IS_RECORDING);
                    answer.putExtra(RECORDING_ID, recordingId);
                    answer.putExtra(IS_RECORDING, isRunning);
                    LocalBroadcastManager.getInstance(this).sendBroadcast(answer);
                    break;
                }
            }
        return START_STICKY;
    }

    private void startDataCollection() {
        this.isRunning = true;
        if (DataCollectorApplication.LOCATION_ENABLED) {
            myLocation = new MyLocation(this);
            foursquareCaller = new FoursquareCaller(this, currentLocation);
            googlePlacesCaller = new GooglePlacesCaller(this);
            prePlaceName = "null";
            currentPlaceName = "null";
            ifLocationChanged = true;
        }
        if (DataCollectorApplication.ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED) {
            myMotion = new MyMotion(this);
        }
        if (DataCollectorApplication.ACTIVITY_ENABLED) {
            myActivity = new MyActivity(this);
            preActivity = new DetectedActivity(DetectedActivity.UNKNOWN, 0);
            currentActivity = new DetectedActivity(DetectedActivity.UNKNOWN, 0);
        }
        if (DataCollectorApplication.ENVIRONMENT_SENSOR_ENABLED) {
            myEnvironmentSensor = new MyEnvironmentSensor(this);
            preAmbientLight = 0;
            currentAmbientLight = 0;
        }
        if (DataCollectorApplication.GOOGLE_FITNESS_ENABLED) {
            googleFitness = new GoogleFitness(this);
            currentSteps = 0;
            preSteps = 0;
        }
        if (DataCollectorApplication.AMBIENT_SOUND_ENABLED) {
            ambientSound = new AmbientSound(this);
            currentAmbientSound = 0;
        }
        if (DataCollectorApplication.WEATHER_ENABLED) {
            weatherCaller = new WeatherCaller(this);
            currentWeatherCondition = "null";
            currentTemperature = 0;
        }

        preWifiName = "null";
        currentWifiName = "null";
        currentLabel = "null";

        LocalBroadcastManager.getInstance(this).registerReceiver(mMessageReceiver,
                new IntentFilter(DataCollectorApplication.BROADCAST_EVENT));

        startScheduledUpdate();
        startNotification();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        isRunning = false;
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        userName = sharedPreferences.getString("fingerprint_user_name", "userName");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        isRunning = false;
        cancelNotification();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mMessageReceiver);

        if (DataCollectorApplication.ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED) {
            stopMotionSensor();
        }
        if (DataCollectorApplication.LOCATION_ENABLED) {
            stopLocationUpdate();
        }
        if (DataCollectorApplication.ACTIVITY_ENABLED) {
            stopActivityDetection();
        }
        if (DataCollectorApplication.ENVIRONMENT_SENSOR_ENABLED) {
            stopEnvironmentSensor();
        }
        if (DataCollectorApplication.GOOGLE_FITNESS_ENABLED) {
            stopGoogleFitness();
        }
        if (ambientSound != null) {
            ambientSound.stop();
            ambientSound = null;
        }
    }

    private void startScheduledUpdate() {
        final Handler handler = new Handler();
        handler.post((new Runnable() {
            public void run() {
                //we assume that the operation finishes before it must be called again
                handler.postDelayed(this, TimeUnit.SECONDS.toMillis(secondsUntilNextUpdate));
                if (isRunning) {
                    if (weatherUpdateCnt == 1) {
                        weatherCaller.getCurrentWeather();
                    } else if (weatherUpdateCnt == 3600 / secondsUntilNextUpdate) {//update
                        // per hour
                        weatherUpdateCnt = 0;
                    }
                    weatherUpdateCnt++;
                    updateUnAutomaticData();
                    uploadDataSet();
                }
            }
        }));
    }

    private void getPlaces() {
        try {
            if (useGooglePlaces) {
                googlePlacesCaller.getCurrentPlace();
            } else {
                foursquareCaller.findPlaces();
            }
        } catch (Exception e) {
            Log.d(TAG, "getPlaces Exception");
        }

    }

    private void updateUnAutomaticData() {
        getWiFiName();
        checkScreenOn();
        ambientSound.getAmbientSound();
        googleFitness.readData();
    }

    private void getWiFiName() {
        try {
            WifiManager wifiMgr = (WifiManager) getApplicationContext().getSystemService(Context
                    .WIFI_SERVICE);
            WifiInfo wifiInfo = wifiMgr.getConnectionInfo();
            String wifi = wifiInfo.getSSID() + ":" + wifiInfo.getBSSID() + ":" + wifiInfo
                    .getIpAddress() + ":" + wifiInfo.getNetworkId() + ":" + wifiInfo.getRssi();
            //String wifi = wifiInfo.getSSID()+":"+wifiInfo.getIpAddress()+":"+wifiInfo
            // .getNetworkId();
            if (wifiInfo.getSSID() == null) {
                currentWifiName = "null";
            } else {
                currentWifiName = wifi;
            }

        } catch (Exception e) {
            Log.d(TAG, "get wifi name error");
        }

    }

    public void checkScreenOn() {
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            DisplayManager dm = (DisplayManager) getApplicationContext().getSystemService(Context
                    .DISPLAY_SERVICE);
            for (Display display : dm.getDisplays()) {
                if (display.getState() == Display.STATE_OFF) {
                    isCurrentScreenOn = false;
                } else {
                    isCurrentScreenOn = true;
                }
            }
        } else {
            PowerManager pm = (PowerManager) getApplicationContext().getSystemService(Context
                    .POWER_SERVICE);
            //noinspection deprecation
            isCurrentScreenOn = pm.isScreenOn();
        }
    }

    float Rotation[] = new float[9];
    float Inclination[] = new float[9];

    private void uploadDataSet() {
        SensorDataSet sensorDataSet = new SensorDataSet(LocalDateTime.now(), "username");
        sensorDataSet.setRecordingId(recordingId);
        if (DataCollectorApplication.ACTIVITY_ENABLED) {
            sensorDataSet.setActivity(currentActivity);
        }
        if (DataCollectorApplication.WIFI_NAME_ENABLED) {
            sensorDataSet.setWifiName(currentWifiName);
        }
        if (DataCollectorApplication.ENVIRONMENT_SENSOR_ENABLED) {
            sensorDataSet.setHumidityPercent(myEnvironmentSensor.humidity);
            sensorDataSet.setAmbientLight(myEnvironmentSensor.light);
            sensorDataSet.setAirPressure(myEnvironmentSensor.pressure);
            sensorDataSet.setTemperature(myEnvironmentSensor.temperature);
        }
        if (DataCollectorApplication.LOCATION_ENABLED) {
            Log.d(TAG, currentLatitude + "");
            Log.d(TAG, currentLongitude + "");
            Log.d(TAG, "" + currentAccurate);
            Location location = new Location("gps");
            location.setLatitude(currentLatitude);
            location.setLongitude(currentLongitude);
            location.setAccuracy((float) currentAccurate);
            sensorDataSet.setGps(location);
            sensorDataSet.setLocation(currentPlaceName);
        }
        if (DataCollectorApplication.GOOGLE_FITNESS_ENABLED) {
            sensorDataSet.setStepsSinceLast(currentSteps);
        }
        if (DataCollectorApplication.AMBIENT_SOUND_ENABLED) {
            sensorDataSet.setAmbientSound(currentAmbientSound);
        }
        if (DataCollectorApplication.WEATHER_ENABLED) {
            sensorDataSet.setWeather(currentWeatherId);
        }
        if (DataCollectorApplication.ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED) {
            float[] avgAcc = averageDequeue(accData);
            sensorDataSet.setAcc_x(avgAcc[0]);
            sensorDataSet.setAcc_y(avgAcc[1]);
            sensorDataSet.setAcc_z(avgAcc[2]);
            float[] avgRot = averageDequeue(rotData);
            sensorDataSet.setGyro_x(avgRot[0]);
            sensorDataSet.setGyro_y(avgRot[1]);
            sensorDataSet.setGyro_z(avgRot[2]);
            float[] avgMag = averageDequeue(magData);
            sensorDataSet.setMag_x(avgMag[0]);
            sensorDataSet.setMag_y(avgMag[1]);
            sensorDataSet.setMag_z(avgMag[2]);
            if (avgAcc != null && avgMag != null) {
                boolean success = SensorManager.getRotationMatrix(Rotation, Inclination, avgAcc,
                        avgMag);
                if (success) {
                    float orientation[] = new float[3];
                    SensorManager.getOrientation(Rotation, orientation);
                    sensorDataSet.setAzimuth(orientation[0]); // orientation contains: azimut,
                    // pitch and roll
                    sensorDataSet.setPitch(orientation[1]);
                    sensorDataSet.setRoll(orientation[2]);
                }
            }
        }
        sensorDataSet.setScreenState(isCurrentScreenOn);
        sensorDataBuffer[sample] = sensorDataSet;
        sample++;
        if (sample >= sensorDataBuffer.length) {
            sample = 0;
        }
        db.insertSensorDataSet(sensorDataSet);
    }

    @Override
    public void activityUpdate(DetectedActivity activity) {
        currentActivity = activity;
    }

    @Override
    public void stopActivityDetection() {
        myActivity.removeActivityUpdates();
        myActivity.disconnect();
        myActivity.unregisterReceiver();
    }

    @Override
    public void environmentSensorDataChanged(float light, float temperature, float pressure,
                                             float humidity) {
        currentAmbientLight = light;
    }

    @Override
    public void stopEnvironmentSensor() {
        myEnvironmentSensor.stopEnvironmentSensor();
    }


    @Override
    public void locationChanged(Location location) {
        if (location != null) {
            currentLocation = location;
            currentLatitude = location.getLatitude();
            currentLongitude = location.getLongitude();
            currentAccurate = location.getAccuracy();
            getPlaces();
        }
    }

    // google places handler
    @Override
    public void onReceivedPlaces(HashMap<String, Float> places) {
        //get the most likely one
        String placeName = null;
        float probability = 0;
        for (Map.Entry<String, Float> pair : places.entrySet()) {
            if ((float) pair.getValue() >= probability) {
                probability = (float) pair.getValue();
                placeName = pair.getKey();
            }
        }

        if (placeName == null) {
            currentPlaceName = "null";
        } else {
            currentPlaceName = placeName + ":" + probability;
        }

        //get all potential places
/*        Iterator it = places.entrySet().iterator();
        String placeName  = "";
        while (it.hasNext()) {
            Map.Entry pair = (Map.Entry)it.next();
            if((float)pair.getValue()>0) {
                placeName += pair.getKey() + " : " + pair.getValue() + ";";
            }
            it.remove(); // avoids a ConcurrentModificationException
        }
        if(placeName.isEmpty()) {
            if(places.entrySet().size()>0){
                placeName = "Unknown Places";
            }else {
                placeName = "null";
            }
        }else {
            currentPlaceName = placeName;
        }*/

    }

    //foursquare place handler
    @Override
    public void placesFound(String place) {
        if (place != null) {
            String[] places = place.split("\n");
            currentPlaceName = places[0];
            return;
        }
        currentPlaceName = "null";
    }

    @Override
    public void stopLocationUpdate() {
        myLocation.stopLocationUpdate();
        googlePlacesCaller.disconnect();
    }


    private ArrayDeque<float[]> accData = new ArrayDeque<>(50);
    private ArrayDeque<float[]> rotData = new ArrayDeque<>(50);
    private ArrayDeque<float[]> magData = new ArrayDeque<>(50);

    @Override
    public void motionDataChanged(float[] accData, float[] rotData, float[] magData) {
        this.accData.add(accData);
        this.rotData.add(rotData);
        this.magData.add(magData);
    }

    @Override
    public void stopMotionSensor() {
        myMotion.stopMotionSensor();
    }

    public void stopGoogleFitness() {
        googleFitness.disconnect();
    }

    @Override
    public void onReceivedAmbientSound(double volume) {
        currentAmbientSound = volume;
    }

    @Override
    public void onReceivedStepsCounter(long steps) {
        currentSteps = steps;
    }

    @Override
    public void onReceivedWeather(String weather) {
        currentWeatherId = db.enterWeather(weather, currentWeatherId + 1);
    }

    public void startNotification() {
        NotificationManager mNotificationManager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationCompat.Builder mNotifyBuilder = new NotificationCompat.Builder(this)
                .setContentTitle("DataCollector")
                .setAutoCancel(false)
                .setOngoing(true)
                .setContentText("Background service is running.")
                .setSmallIcon(R.drawable.fp_s).addAction(0, "10 minutes", PendingIntent
                        .getService(this, 19921, new Intent(this, ActivitiesIntentService.class)
                                .putExtra(TYPE, MINUTES), PendingIntent.FLAG_ONE_SHOT))
                .addAction(0, "activity", PendingIntent.getService(this, 19922, new Intent(this,
                        ActivitiesIntentService.class).putExtra(TYPE, ACTIVITY), PendingIntent
                        .FLAG_ONE_SHOT))
                .addAction(0, "snack", PendingIntent.getService(this, 19923, new Intent(this,
                        ActivitiesIntentService.class).putExtra(TYPE, SNACK), PendingIntent
                        .FLAG_ONE_SHOT));


        mNotificationManager.notify(
                notificationID,
                mNotifyBuilder.build());
    }

    public void cancelNotification() {
        NotificationManager mNotificationManager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.cancel(notificationID);
    }

    private boolean checkChange() {
        int change = 0;
        if (!currentActivity.equals(preActivity)) {
            preActivity = currentActivity;
            change++;
        }
        if (currentSteps != preSteps) {
            preSteps = currentSteps;
            change++;
        }
        if (!currentWifiName.equals(preWifiName)) {
            preWifiName = currentWifiName;
            change++;
        }
        if (!currentPlaceName.equals(prePlaceName)) {
            prePlaceName = currentPlaceName;
            change++;
        }
        if (!isCurrentScreenOn == isPreScreenOn) {
            isPreScreenOn = isCurrentScreenOn;
            change++;
        }
        if (change > 0) {
            return true;
        }
        return false;
    }

    private BroadcastReceiver mMessageReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            currentLabel = intent.getStringExtra("label");
            //Log.d("receiver", "Got message: " + message);
        }
    };
}

File: datacollector/src/main/java/de/leo/smartTrigger/datacollector/datacollection/DataCollectorService.kt
package de.leo.smartTrigger.datacollector.datacollection

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.location.Location
import android.net.wifi.WifiManager
import android.os.Handler
import android.os.IBinder
import android.os.PowerManager
import android.preference.PreferenceManager
import android.util.Log
import androidx.core.app.NotificationCompat
import com.google.android.gms.location.*
import de.leo.smartTrigger.datacollector.R
import de.leo.smartTrigger.datacollector.datacollection.database.*
import de.leo.smartTrigger.datacollector.datacollection.models.SensorDataSet
import de.leo.smartTrigger.datacollector.datacollection.models.WifiInfo
import de.leo.smartTrigger.datacollector.datacollection.sensors.*
import de.leo.smartTrigger.datacollector.jitai.UNKNOWN_LOCATION
import de.leo.smartTrigger.datacollector.jitai.activityDetection.ActivityRecognizer
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.ACTIVITY_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.AMBIENT_SOUND_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.ENVIRONMENT_SENSOR_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.GOOGLE_FITNESS_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.LOCATION_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.LOCATION_NAME_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.SCREEN_ON_ENABLED
import de.leo.smartTrigger.datacollector.ui.application.DataCollectorApplication.Companion.WEATHER_ENABLED
import de.leo.smartTrigger.datacollector.ui.naturalTrigger.list.TriggerManagingActivity
import de.leo.smartTrigger.datacollector.ui.notifications.NotificationService.Companion.CHANNEL
import de.leo.smartTrigger.datacollector.utils.UPDATE_JITAI
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.intentFor
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeUnit.MINUTES
import java.util.concurrent.TimeUnit.SECONDS
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class DataCollectorService : Service(),
                             FourSquareListener,
                             MyActivityListener,
                             GooglePlacesListener,
                             AmbientSoundListener,
                             GoogleFitnessListener,
                             WeatherCallerListener {


    companion object {

        val TAG = "DataCollectorService"

        val notificationID = 1001
        val UPDATE_DELAY = SECONDS.toMillis(10L)
        private val WEATHER_UPDATE_DELAY = MINUTES.toMillis(30)

        private val useGooglePlaces = false
    }

    private var weatherUpdateCnt = 0L

    private lateinit var myMotion: MotionSensors

    private lateinit var myActivity: MyActivity

    private var googlePlacesCaller: GooglePlacesCaller? = null
    private var currentLocation: Location = Location(UNKNOWN_LOCATION)

    private var myEnvironmentSensor: MyEnvironmentSensor? = null

    private var ambientSound: AmbientSound? = null
    private lateinit var googleFitness: GoogleFitness
    private lateinit var weatherCaller: WeatherCaller

    private var currentAmbientSound: Double = 0.0
    private var isRunning: Boolean = false
    private var currentPlaceName: String = ""
    private var currentActivities = listOf(DetectedActivity(DetectedActivity.UNKNOWN, 0))
    private var currentWeatherId: Long = -1
    private var currentSteps: Long = 0
    private val userName: String by lazy {
        PreferenceManager.getDefaultSharedPreferences(this)
            .getString(getString(R.string.user_name), null)
    }
    private var db: JitaiDatabase? = null
    private lateinit var activityRecognizer: ActivityRecognizer
    private lateinit var wifiScanner: WifiScanner

    private val pm: PowerManager by lazy { applicationContext.getSystemService(Context.POWER_SERVICE) as PowerManager }

    private var updateDelay: Long = UPDATE_DELAY
        set(value) {
            if (field != value) {
                field = value
                startNotification(field)
            }
        }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (!this::wifiScanner.isInitialized)
            wifiScanner = WifiScanner(this)
        db = JitaiDatabase.getInstance(applicationContext)
        currentWeatherId = db!!.getLatestWeather()!!.id.toLong()
        Log.d(TAG, "service recieved ${intent?.action ?: "unknown signal"}")
        if (!this.isRunning) {
            startDataCollection()
        } else {
            //reschedule to update updatedelay
            startScheduledUpdate()
        }
        if (intent != null && intent.action != null)
            when (intent.action) {
                UPDATE_JITAI -> {
                    val id = intent.getIntExtra(JITAI_ID, -1)
                    if (id != -1)
                        activityRecognizer.updateNaturalTrigger(id)
                    else
                        activityRecognizer = ActivityRecognizer(baseContext)
                }
            }
        return Service.START_STICKY
    }

    private fun startDataCollection() {
        activityRecognizer = ActivityRecognizer(baseContext)
        this.isRunning = true
        if (LOCATION_NAME_ENABLED) {
            googlePlacesCaller = GooglePlacesCaller(this)
        }
        if (ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED) {
            myMotion = MotionSensors(this)
        }
        if (ACTIVITY_ENABLED) {
            myActivity = MyActivity(this)
            currentActivities = listOf(DetectedActivity(DetectedActivity.UNKNOWN, 0))
        }
        if (ENVIRONMENT_SENSOR_ENABLED) {
            myEnvironmentSensor = MyEnvironmentSensor(this)
        }
        if (GOOGLE_FITNESS_ENABLED) {
            googleFitness = GoogleFitness(this)
            currentSteps = 0
        }
        if (AMBIENT_SOUND_ENABLED) {
            ambientSound = AmbientSound(this)
            currentAmbientSound = 0.0
        }
        if (WEATHER_ENABLED) {
            weatherCaller = WeatherCaller(this)
        }
        startScheduledUpdate()
    }

    override fun onCreate() {
        super.onCreate()
        isRunning = false
        startNotification(updateDelay)
    }

    override fun onDestroy() {
        super.onDestroy()
        isRunning = false
        cancelNotification()

        if (DataCollectorApplication.ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED) {
            myMotion.stopMotionSensor()
        }
        if (DataCollectorApplication.ACTIVITY_ENABLED) {
            stopActivityDetection()
        }
        if (DataCollectorApplication.ENVIRONMENT_SENSOR_ENABLED) {
            myEnvironmentSensor?.stopEnvironmentSensor()
        }
        if (DataCollectorApplication.GOOGLE_FITNESS_ENABLED) {
            stopGoogleFitness()
        }
        ambientSound?.stop()
        ambientSound = null
    }

    private fun startScheduledUpdate() {
        val handler = Handler()
        handler.post(object : Runnable {
            override fun run() {
                val currentTime = System.currentTimeMillis()
                if (isRunning) {
                    uploadDataSet(currentTime)
                }
                updateDelay = Math.max(
                    Math.min(activityRecognizer.nextUpdate(), TimeUnit.MINUTES.toMillis(5)),
                    UPDATE_DELAY)
                handler.removeCallbacks { this }
                handler.postDelayed(this, updateDelay)
            }
        })
    }

    private val wifiMgr by lazy {
        applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
    }

    suspend fun getWifi(): List<WifiInfo>? {
        val currentWifis = wifiScanner.getCurrentScanResult()
        var wifiInfo: android.net.wifi.WifiInfo? = null
        try {
            wifiInfo = wifiMgr.connectionInfo
        } catch (e: Exception) {
            Log.d(TAG, "get wifi name error")
        }
        if (currentWifis.isEmpty()) {
            return wifiInfo?.let {
                if (wifiInfo.bssid != null)
                    listOf(WifiInfo(wifiInfo))
                else null
            }
        } else {
            return currentWifis.map {
                //replace the connected wifi with better information
                if (wifiInfo?.bssid == it.BSSID) {
                    WifiInfo(wifiInfo!!)
                } else {
                    WifiInfo(it)
                }
            }
        }
    }


    fun checkScreenOn(): Boolean = pm.isInteractive

    private fun uploadDataSet(currentTime: Long) {
        //trigger asynchronus processes
        if (WEATHER_ENABLED) {
            if (weatherUpdateCnt == WEATHER_UPDATE_DELAY) {
                weatherCaller.getCurrentWeather()
                weatherUpdateCnt = 0
            }
            weatherUpdateCnt++
        }
        if (AMBIENT_SOUND_ENABLED) ambientSound?.getAmbientSound()
        if (GOOGLE_FITNESS_ENABLED) googleFitness.readData()
        //retrieve most recent data and store
        val sensorDataSet = SensorDataSet(currentTime, userName)
        if (ACTIVITY_ENABLED) sensorDataSet.activity = currentActivities
        if (GOOGLE_FITNESS_ENABLED) {
            sensorDataSet.totalStepsToday = currentSteps
        }
        if (AMBIENT_SOUND_ENABLED) {
            sensorDataSet.ambientSound = currentAmbientSound
        }
        if (WEATHER_ENABLED) {
            sensorDataSet.weather = currentWeatherId
        }
        if (SCREEN_ON_ENABLED) sensorDataSet.screenState = checkScreenOn()
        GlobalScope.launch(Dispatchers.Main) {
            val location = async { getLocation() }
            val wifi = async { getWifi() }
            currentLocation = location.await()
            val currentWifiInfo = wifi.await()
            Log.d(TAG, "lat: ${currentLocation.latitude}," +
                " long: ${currentLocation.longitude}," +
                " accuracy ${currentLocation.accuracy}")
            sensorDataSet.gps = currentLocation
            sensorDataSet.wifiInformation = currentWifiInfo
            sensorDataSet.locationName = currentPlaceName
            sensorDataSet.id = db!!.insertSensorDataSet(sensorDataSet)
            activityRecognizer.recognizeActivity(sensorDataSet)
            sensorDataSet.id = db!!.insertSensorDataSet(sensorDataSet)
            activityRecognizer.recognizeActivity(sensorDataSet)
        }
        uploadMotionData()
        uploadEnvironmentData()
    }

    suspend fun getLocation(): Location =
        if (LOCATION_ENABLED)
            getCurrentLocation(this) ?: Location(UNKNOWN_LOCATION)
        else
            Location(UNKNOWN_LOCATION)

    private fun uploadEnvironmentData() {
        if (DataCollectorApplication.ENVIRONMENT_SENSOR_ENABLED) {
            doAsync {
                db!!.enterSingleDimensionDataBatch(TABLE_REALTIME_AIR,
                                                   myEnvironmentSensor!!.readPressureData())
                db!!.enterSingleDimensionDataBatch(TABLE_REALTIME_TEMPERATURE,
                                                   myEnvironmentSensor!!.readTemperatureData())
                db!!.enterSingleDimensionDataBatch(TABLE_REALTIME_LIGHT,
                                                   myEnvironmentSensor!!.readLightData())
                db!!.enterSingleDimensionDataBatch(TABLE_REALTIME_HUMIDITY,
                                                   myEnvironmentSensor!!.readHumidityData())
                db!!.enterSingleDimensionDataBatch(TABLE_REALTIME_PROXIMITY,
                                                   myEnvironmentSensor!!.readProximityData())
            }
        }
    }

    private fun uploadMotionData() {
        if (DataCollectorApplication.ACCELEROMETER_MAGNETOMETER_GYROSCOPE_ORIENTATION_ENABLED) {
            val accData = myMotion.readAccData()
            val rotData = myMotion.readRotData()
            val magData = myMotion.readMagData()
            val gyroData = myMotion.readGyroData()
            doAsync {
                db!!.enterAccDataBatch(accData)
                db!!.enterGyroDataBatch(gyroData)
                db!!.enterMagDataBatch(magData)
                db!!.enterRotDataBatch(rotData)
            }
        }
    }

    override fun activityUpdate(activity: List<DetectedActivity>) {
        currentActivities = activity
    }

    override fun stopActivityDetection() {
        myActivity.removeActivityUpdates()
        myActivity.disconnect()
        myActivity.unregisterReceiver()
    }

    // google places handler
    override fun onReceivedPlaces(places: HashMap<String, Float>) {
        //get the most likely one
        var placeName: String? = null
        var probability = 0f
        for ((key, value) in places) {
            if (value >= probability) {
                probability = value
                placeName = key
            }
        }

        if (placeName == null) {
            currentPlaceName = "null"
        } else {
            currentPlaceName = placeName + ":" + probability
        }
    }

    @SuppressLint("MissingPermission")
    suspend fun getCurrentLocation(context: Context): Location? = suspendCoroutine { cont ->
        val fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
        val locationRequest = LocationRequest()
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        locationRequest.numUpdates = 1
        val listener = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult?) {
                cont.resume(locationResult?.lastLocation)
            }
        }
        val task = fusedLocationProviderClient.requestLocationUpdates(locationRequest,
                                                                      listener,
                                                                      null)
        task.addOnFailureListener { cont.resumeWithException(it) }
    }

    //foursquare place handler
    override fun placesFound(place: String?) {
        if (place != null) {
            val places = place.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            currentPlaceName = places[0]
            return
        }
        currentPlaceName = "null"
    }

    fun stopGoogleFitness() {
        googleFitness.disconnect()
    }

    override fun onReceivedAmbientSound(volume: Double) {
        currentAmbientSound = volume
    }

    override fun onReceivedStepsCounter(steps: Long) {
        currentSteps = steps
    }

    override fun onReceivedWeather(weather: String?) {
        if (weather != null)
            currentWeatherId = db!!.enterWeather(weather, currentWeatherId + 1)
    }

    fun startNotification(updateDelay: Long) {
        val contentIntent = intentFor<TriggerManagingActivity>().setAction("OPEN_TRIGGER_LIST")
        val mNotifyBuilder = NotificationCompat.Builder(this, CHANNEL)
            .setContentIntent(PendingIntent.getActivity(this,
                                                        6874,
                                                        contentIntent,
                                                        PendingIntent.FLAG_UPDATE_CURRENT))
            .setContentTitle("Smart Trigger")
            .setAutoCancel(false)
            .setOngoing(true)
            .setContentText("Studienanwendung läuft."// Current rate: ${TimeUnit.MILLISECONDS
                // .toSeconds(updateDelay)}s"
                           )
            .setSmallIcon(R.drawable.ic_reminder_notification)
            .setColor(resources.getColor(R.color.green_800))
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setOnlyAlertOnce(true)
        startForeground(notificationID, mNotifyBuilder.build())
    }

    fun cancelNotification() {
        stopForeground(true)
    }
}
